### DISCUSSION TITLE
PyCharm setup for rmf_core

### DISCUSSION THREAD

**Discussion by albertleng** [osrf/rmf_core#244] [Q&A]
I am developing a fleet adapter based on `rmf_core`  in PyCharm. I want to have autocomplete etc. to ease my development. I followed steps in [PyCharm import external library](https://stackoverflow.com/a/36492328) and [IDEs](http://wiki.ros.org/IDEs) and managed to setup the environment for ROS/ROS2.  I also added the path to `rmf_core` into Interpreter Paths as shown in picture below.
![PyCharm_ProjectInterpreter_Settings02](https://user-images.githubusercontent.com/24582966/103504287-f0140380-4e91-11eb-8fac-c86f82d3fec5.png)

However, I still get the warning red curvy lines for `FleetState`, `RobotState` etc as shown below.  

Any idea how to resolve that? Thanks.

![rmf_messages_warning02](https://user-images.githubusercontent.com/24582966/103504413-4d0fb980-4e92-11eb-903d-30eac6088a03.png)

**Comment by mxgrey** [comment]
I think you'll want to seek advice from [pybind11](https://github.com/pybind/pybind11) which is the tool used to create the python bindings.

But also please note that the python bindings are provided in the [rmf_fleet_adapter_python](https://github.com/osrf/rmf_fleet_adapter_python) repository, not in `rmf_core`. I think you'll want to build `rmf_fleet_adapter_python` and then source the python library that gets produced from building that package.

**Reply by albertleng** [reply]
Thanks @mxgrey for your answer.  

I think my issue is not related to python binding per se. After adding the path (as shown in the screenshot) to interpreter paths, the red curvy lines for `FleetState`,  `RobotState` etc. go away.  

![rmf_fleet_msgs](https://user-images.githubusercontent.com/24582966/103604802-1a76c700-4f4d-11eb-9b0f-79809af2a51a.png)

**Reply by albertleng** [reply]
> I think you'll want to seek advice from [pybind11](https://github.com/pybind/pybind11) which is the tool used to create the python bindings.
> 
> But also please note that the python bindings are provided in the [rmf_fleet_adapter_python](https://github.com/osrf/rmf_fleet_adapter_python) repository, not in `rmf_core`. I think you'll want to build `rmf_fleet_adapter_python` and then source the python library that gets produced from building that package.

On a sidenote, I wanted to tinker with `rmf_fleet_adapter_python`. Following the instructions in  [rmf_fleet_adapter_python](https://github.com/osrf/rmf_fleet_adapter_python) repository, building `rmf_fleet_adapter_python` via `colcon build --packages-select rmf_fleet_adapter_python` gave me the following errors:  
``` bash
Starting >>> rmf_fleet_adapter_python
--- stderr: rmf_fleet_adapter_python                   
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find FastRTPS (missing: FastRTPS_INCLUDE_DIR FastRTPS_LIBRARIES)
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /opt/ros/foxy/share/fastrtps_cmake_module/cmake/Modules/FindFastRTPS.cmake:95 (find_package_handle_standard_args)
  /opt/ros/foxy/share/rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_c-extras.cmake:7 (find_package)
  /opt/ros/foxy/share/rosidl_typesupport_fastrtps_c/cmake/rosidl_typesupport_fastrtps_cConfig.cmake:41 (include)
  /opt/ros/foxy/share/rosidl_typesupport_c/cmake/rosidl_typesupport_c-extras.cmake:13 (find_package)
  /opt/ros/foxy/share/rosidl_typesupport_c/cmake/rosidl_typesupport_cConfig.cmake:41 (include)
  /opt/ros/foxy/share/rosidl_default_runtime/cmake/rosidl_default_runtime-extras.cmake:12 (find_package)
  /opt/ros/foxy/share/rosidl_default_runtime/cmake/rosidl_default_runtimeConfig.cmake:41 (include)
  /opt/ros/foxy/share/builtin_interfaces/cmake/ament_cmake_export_dependencies-extras.cmake:21 (find_package)
  /opt/ros/foxy/share/builtin_interfaces/cmake/builtin_interfacesConfig.cmake:41 (include)
  /home/albert/ws_rmf/install/rmf_traffic_msgs/share/rmf_traffic_msgs/cmake/ament_cmake_export_dependencies-extras.cmake:21 (find_package)
  /home/albert/ws_rmf/install/rmf_traffic_msgs/share/rmf_traffic_msgs/cmake/rmf_traffic_msgsConfig.cmake:41 (include)
  /home/albert/ws_rmf/install/rmf_traffic_ros2/share/rmf_traffic_ros2/cmake/ament_cmake_export_dependencies-extras.cmake:21 (find_package)
  /home/albert/ws_rmf/install/rmf_traffic_ros2/share/rmf_traffic_ros2/cmake/rmf_traffic_ros2Config.cmake:41 (include)
  /home/albert/ws_rmf/install/rmf_fleet_adapter/share/rmf_fleet_adapter/cmake/ament_cmake_export_dependencies-extras.cmake:21 (find_package)
  /home/albert/ws_rmf/install/rmf_fleet_adapter/share/rmf_fleet_adapter/cmake/rmf_fleet_adapterConfig.cmake:41 (include)
  CMakeLists.txt:22 (find_package)


Traceback (most recent call last):
  File "setup.py", line 77, in <module>
    setup(
  File "/usr/lib/python3/dist-packages/setuptools/__init__.py", line 144, in setup
    return distutils.core.setup(**attrs)
  File "/usr/lib/python3.8/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib/python3.8/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/usr/lib/python3.8/distutils/command/build.py", line 135, in run
    self.run_command(cmd_name)
  File "/usr/lib/python3.8/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "setup.py", line 37, in run
    self.build_extension(ext)
  File "setup.py", line 70, in build_extension
    subprocess.check_call(['cmake', ext.sourcedir] + cmake_args,
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '/home/albert/ws_rmf/src/rmf_fleet_adapter_python', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/home/albert/ws_rmf/build/rmf_fleet_adapter_python/build/lib.linux-x86_64-3.8/', '-DPYTHON_EXECUTABLE=/usr/bin/python3', '-DCMAKE_BUILD_TYPE=Release']' returned non-zero exit status 1.
---
Failed   <<< rmf_fleet_adapter_python [1.23s, exited with code 1]

Summary: 0 packages finished [1.37s]
  1 package failed: rmf_fleet_adapter_python
  1 package had stderr output: rmf_fleet_adapter_python

```
I have sourced the workspace. There are no tabs for `discussions` and `issues` in  [rmf_fleet_adapter_python](https://github.com/osrf/rmf_fleet_adapter_python) repository. Hence, I am seeking for help here.

**Comment by mxgrey** [comment]
> There are no tabs for discussions and issues in rmf_fleet_adapter_python repository. Hence, I am seeking for help here.

Thanks for pointing that out; that was an oversight. Those tabs should be activated now. I've activated the `Issues` tab for that repo, but we can continue using the discussion board of this repo for all matters related to RMF.

> gave me the following errors:

This is a known issue in the Foxy distribution which should have been fixed in a recent sync. If you haven't updated your foxy distro in a few weeks, then I recommend trying:

```
$ sudo apt update
$ sudo apt dist-upgrade
```

If that still fails for some reason, then a sure-fire way to fix that error is to clone the FastRTPS repo into your workspace.
